include/HAL/TargetSystem.h

```diff
@@ -139,6 +141,31 @@ class Target {
   /// @brief Disable(stop) ongoing timers
   void disableTiming();
 
+  // Diagnostic access
+  /// @brief Add a diagnostic to this target, not thread safe
+  void addDiagnostic(const qssc::Diagnostic &diag) {
+    diagnostics_.emplace_back(diag);
+  }
+  /// @brief Add a diagnostic to this target using mutex locks for thread safety
+  void addDiagnosticWithLock(const qssc::Diagnostic &diag) {
+    const std::lock_guard<std::mutex> lock(diagnosticsMutex_);
+    diagnostics_.emplace_back(diag);
+  }
+  /// @brief Get the diagnostics that this target holds
+  const qssc::DiagList &getDiagnosticList() const { return diagnostics_; }
```

Comment:
Does this pop the diagnostics? If not how do we clear the list of diagnostics?

Reply:
The diagnostics will be destroyed when the target is destroyed. Currently the target copies the diagnostic that is given to it by the `addDiagnostic` calls. I did this because the diagnostic created by the scheduler is owned by the command there and couldn't be moved to give it to the target without some rework there. `getDiagnosticList` just hands a const reference to the internal target's list. The external API is really focused on `getDiagnosticsRecursive`, which creates and returns a new list containing references to all of the target tree's held diagnostics.

Reply:
I think we should refactor this a bit more so that we fetch diagnostics after each target invocation. That way we can localize errors to the respective target call.

Reply:
Diagnostics are now removed and returned in https://github.com/openqasm/qe-compiler/pull/240/commits/8e9e83e4ff24caeb8bb623da61eddf8e56a961a8